pkg/pod/pod.go

```diff
@@ -98,6 +116,58 @@ func (c *Client) GetPods() (pods []*v1.Pod, err error) {
 	return listPods, nil
 }
 
+// ListPods returns list of all pods
+func (c *Client) ListPods() (pods []*v1.Pod, err error) {
+	var resList []*v1.Pod
+	listPods := c.PodWatcher.Informer().GetStore().List()
+
+	for _, pod := range listPods {
+		v1Pod, ok := pod.(*v1.Pod)
+		if !ok {
+			err := fmt.Errorf("could not cast %T to v1.Pod", pod)
+			klog.Error(err)
+			return nil, err
+		}
+
+		resList = append(resList, v1Pod)
+		klog.V(6).Infof("Appending pod ID: %s/%s to list.", v1Pod.UID, v1Pod.Name)
+	}
+
+	return resList, nil
+}
+
+// GetPod returns the pod object
+func GetPod(obj interface{}, i informersv1.PodInformer) (pod *v1.Pod, err error) {
+
+	tries := 0
+	totalRetries := 300
```

Comment:
>totalRetries := 300 [](start = 1, length = 19)

We need to revise this logic - with 300 retries and explicit 1 second sleeps between them I am pretty sure we are using APIs in a wrong way. The fact that this call can take up to 5 minutes is definitely wrong way to use this API.

Reply:
AFAIK the "informer" pattern watches for object changes and notifies clients about those changes. Getting data store to which informer is connected and then starting explicitly polling for it looks like a bruteforce hack

---
In reply to: [383479968](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r383479968) [](ancestors = 383479968)

Reply:
yes, I agree that it looks like a bruteforce hack. The reason for this is once the informer  detects a pod is created, it does not get the new pod's ip info: Pod.status.PodIP. Though wait for a while, the obj.(*v1.Pod) still does not return the Status.PodIP. So use  another call i.Informer().GetStore().Get(obj) to get its IP info. Let me say if I can so something like refresh to get ip info which can avoid the Getpod function here.

Reply:
I see those Sleep() calls survived to the last iteration. These are potential time bombs that can cause all sorts of timing and performance issues in a busy working cluster.

Reply:
yes. You are absolutely right!!!!!! I never need to call the getpod function. Just check in the UpdateFunc is good enough. Will update the code. Thanks a lot for the great comments!

Reply:
I implement it by checking the ip change for the pods. Let me know if you have any comments.